### ISSUE TITLE
Guidance on initializable iterators w/ numpy arrays

### ISSUE THREAD

**Issue by jperl** [tensorflow/io#138]
I am collecting data during the training process and using Dataset.from_tensor_slices with placeholders and an initializable iterator to refresh the dataset. The dataset uses the tensor slices to then do further preprocessing.

As new data is collected, I reinitialize the iterator's placeholders with the new numpy array data.

Since initializable iterators are deprecated now, how do you recommend I seed the dataset with the dynamics numpy arrays? Should I switch to using a generator?

**Comment by jperl** [comment]
Relates to https://github.com/tensorflow/io/issues/68

**Comment by suphoff** [comment]
It looks like the Iterator class even completely disappeared from the r2.0 API documentation.
As such I would concentrate on creating a dynamic modifiable dataset pipeline. This could be a custom dataset op - or a dataset pipeline using TF variables.
For a good example for the later - take a look at the replay buffer implemented in tensorflow/agent:
https://github.com/tensorflow/agents/blob/master/tf_agents/replay_buffers/tf_uniform_replay_buffer.py
https://github.com/tensorflow/agents/blob/master/tf_agents/replay_buffers/table.py

**Comment by jperl** [comment]
Using variables is a good idea, thank you!

**Comment by yongtang** [comment]
@jperl I think it depends on if your data is already in memory (numpy arrays) or not. If the data is already in numpy arrays then I think we could help implement a Dataset to map the numpy arrays into tensors as data continue to feed in. If you have other sources of input (like a file or a stream), we could also help implement a Dataset to take the input directly, not even necessarily to read into numpy array.

Maybe you could share some details or some boilerplate code to show what the data input format is?

**Comment by jperl** [comment]
The data is paths to images stored on a local filesystem. It is queried from a database, and then in memory as an array of strings. After each training loop we reinitialize the dataset with the latest paths (and other metadata).

I believe we can accomplish this similar to the `TFUniformReplayBuffer` example from @suphoff, using a Counter, map and variables or a generator to provide the latest numpy arrays.
